### ISSUE TITLE
Tutorial 3: Off-policy training with DQN

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#309]
Child of #303 (tutorial series). Add `docs/tutorials/03-dqn.md`: DQN (Double-DQN) on CartPole/GridWorld — replay buffer, target network, ε-annealing, Polyak soft updates, and *when to prefer off-policy DQN over on-policy PPO*.

**Requirements (match tutorials 1–2):**
- Single Markdown file with complete copy-paste code.
- Code CI-enforced as a doc-test via the `src/tutorials.rs` `#[doc = include_str!(...)]` mechanism landed in #303. Add a `tutorial_03_dqn` submodule.
- Voice matches GETTING_STARTED.md; explain *why* at each step.
- Prerequisite framing: builds on tutorial 2's on-policy loop as a contrast.
- Update `docs/tutorials/README.md` status row to ✅ Landed.
- Full test suite + clippy -D warnings + fmt + `RUSTDOCFLAGS=-D warnings cargo doc` clean.

Reference the packaged `train_cartpole_dqn` / `train_dqn_grid_world` examples for the working API.

**Comment by rjwalters** [comment]
Curating this issue per user request (has `loom:triage`). Adding technical context, acceptance criteria, implementation guidance, and test plan grounded in the current codebase on `origin/main`.

---

## Technical Context

### What already exists on `origin/main`

**Tutorial infrastructure (landed in #303):**

- `docs/tutorials/01-your-first-agent.md` — 270 lines; bandit + actor-critic by hand
- `docs/tutorials/02-cartpole-ppo.md` — 320 lines; CartPole + PPO + GAE; hands the loop to `PPOTrainerBurn`
- `docs/tutorials/README.md` — dependency-ordered index; tutorial 3 row already references this issue as "🚧 Planned"
- `src/tutorials.rs` — the doc-test harness; uses `#[doc = include_str!("../docs/tutorials/NN-name.md")]` to pull each Markdown file's `rust` blocks in as doc-tests; guarded by `#[cfg(feature = "training")]`
- `src/lib.rs` at line 149 — `pub mod tutorials;` (already declared, feature-gated)

**DQN API (all on `origin/main`):**

- `src/policy/q_network.rs` — `QNetworkBurn<B>` (2-layer Tanh MLP, orthogonal init). Constructor: `QNetworkBurn::with_seed(obs_dim, n_actions, hidden_dim, seed, &device)`. Forward: `q.forward(obs_tensor)` → `[batch, n_actions]`.
- `src/train/dqn/config.rs` — `DQNConfig` builder. Every field has a setter method. Key setters: `.learning_rate()`, `.batch_size()`, `.buffer_capacity()`, `.min_buffer_size()`, `.target_update_interval()`, `.gamma()`, `.epsilon_start()`, `.epsilon_end()`, `.epsilon_decay_steps()`, `.max_grad_norm()`, `.soft_update_tau(tau)` (enables Polyak instead of hard copy).
- `src/train/dqn/trainer.rs` — `DQNTrainerBurn<B, Q, O>`. Key methods: `::new(config, online, burn_opt, obs_dim, n_actions, device)`, `.select_action(&obs, &mut rng, greedy_fn)`, `.buffer_mut().push(&obs, action, reward, &next_obs, done)`, `.increment_env_step()`, `.maybe_sync_target(blend_fn)`, `.train_step(&mut rng, online_fwd, target_fwd)`.
- `src/train/dqn/loss.rs` — `compute_td_target_double` implements Double-DQN: `a* = argmax Q_online(s')`, then `y = r + γ(1-done)·Q_target(s', a*)`. The trainer uses this unconditionally.
- `src/buffer/replay/storage.rs` — `ReplayBuffer`: fixed-capacity FIFO ring buffer over flat `Vec`s; `push` overwrites oldest on full.

**Reference examples (working, runnable):**

- `examples/games/cartpole/train_cartpole_dqn.rs` — 60k-step CartPole run. Config: `buffer_capacity(50_000)`, `min_buffer_size(1_000)`, `epsilon_start(1.0) → epsilon_end(0.05)` over `epsilon_decay_steps(10_000)`, `soft_update_tau(0.005)`, `target_update_interval(500)`, `gamma(0.99)`. Seeded via `QNetworkBurn::with_seed(..., SEED=0, ...)` + `StdRng::seed_from_u64(0xC0FFEE)`. Supports `CURVE_CSV` + `TOTAL_TIMESTEPS` env vars.
- `examples/games/grid_world/train_dqn_grid_world.rs` — 80k-step GridWorld run (sparse reward: +1.0 at goal, -1.0 in hole, -0.01/step). Obs: 16-dim one-hot. Config: `gamma(0.95)`, `epsilon_end(0.10)`, `epsilon_decay_steps(20_000)`, `batch_size(128)`.

**Existing integration test:**

- `tests/test_dqn_grid_world.rs` — two-tier: fast smoke test (always runs) + `#[ignore]` convergence bar (mean eval return ≥ +0.90 over 20 greedy episodes, within 80k steps).

### How tutorials 1–2 structure their single `rust` block

Each tutorial has exactly one `rust` block. The boilerplate wrapper that makes it a doc-test compilable under `--features training`:

```rust
# #[cfg(feature = "training")]
# fn main() {
// ... full self-contained code ...
# }
# #[cfg(not(feature = "training"))]
# fn main() {}
```

The `#`-prefixed lines are hidden in rendered docs but present in the doc-test runner. Tutorial 2's code block is ~170 lines including the `compute_gae` helper at the bottom.

### Section structure to match (tutorials 1–2 pattern)

```
# Tutorial 3 — Off-policy training with DQN
> callout: You will / Prerequisites / Time
prose intro: how this contrasts with tutorial 2's on-policy loop
## The problem: CartPole-v1 (and a second env, GridWorld, briefly)
## Why off-policy?  (replay buffer breaks correlation; can reuse transitions)
## The four pieces   (Q-network, replay buffer, target network, ε-schedule)
## Double-DQN, briefly
## Polyak soft updates vs hard copy
## The code          (single rust block, CI-enforced)
## Reading the learning curve
## Try it yourself   (3–4 experiments)
## Next              (Tutorial 4: SAC / continuous control)
```

---

## Acceptance Criteria

- [ ] `docs/tutorials/03-dqn.md` exists as a single Markdown file following the Tutorial 1–2 section structure (title, callout header block, one `rust` doc-test block, curve-reading section, "Try it yourself", "Next" pointer).
- [ ] The `rust` block is complete and copy-paste runnable: it imports from `thrust_rl::prelude::*` or the concrete paths, uses `QNetworkBurn::with_seed`, `DQNConfig` builder, `DQNTrainerBurn`, `BurnOptimizer`, and runs at least one `trainer.train_step()` call.
- [ ] The tutorial explains *why* at each step: why replay breaks correlations, why the target network stabilizes training, why ε must anneal rather than stay fixed, why Double-DQN reduces overestimation, and when to reach for DQN over PPO.
- [ ] Prerequisite framing: the callout block lists Tutorial 2 as prerequisite; the prose intro explicitly contrasts the off-policy replay/target-network loop with Tutorial 2's on-policy rollout → GAE → `PPOTrainerBurn` loop.
- [ ] Voice matches GETTING_STARTED.md and tutorials 1–2: second-person, explains concepts before code, comments in-code explain *why* not just *what*.
- [ ] `src/tutorials.rs` gains a new `tutorial_03_dqn` submodule:
  ```rust
  /// Tutorial 3 — Off-policy training with DQN.
  ///
  /// See the rendered Markdown at
  /// [`docs/tutorials/03-dqn.md`](...).
  #[doc = include_str!("../docs/tutorials/03-dqn.md")]
  pub mod tutorial_03_dqn {}
  ```
  The module-level doc comment on `tutorials.rs` must also list `tutorial_03_dqn` in its "Landed tutorials" bullet list.
- [ ] `docs/tutorials/README.md` status row for tutorial 3 changes from `🚧 Planned ([#309])` to `✅ Landed` (issue link removed, link to the `.md` file added).
- [ ] CI gate passes clean:
  - `cargo test --features training` (doc-tests in `tutorial_03_dqn` module compile and run)
  - `cargo clippy --features training -- -D warnings`
  - `cargo fmt --check`
  - `RUSTDOCFLAGS=-D warnings cargo doc --features training`

---

## Implementation Guidance

### Step 1 — Write `docs/tutorials/03-dqn.md`

Model the file on `02-cartpole-ppo.md` (320 lines). The single `rust` block should:

- Keep `TOTAL_TIMESTEPS` small (≤ 8_192) for fast CI; include a comment noting how to bump it for a real run.
- Use CartPole as the primary env (same as tutorial 2, makes the PPO vs DQN contrast concrete).
- Briefly show how to swap to `GridWorld` in a "Try it yourself" exercise rather than embedding two full loops.
- Mirror the CartPole DQN example's config exactly so CI runs reproduce the reference example's behavior, just at a shorter budget.

Key imports the code block needs (drawn from the working examples):
```rust
use burn::{
    backend::{Autodiff, NdArray},
    optim::AdamConfig,
    tensor::{Tensor, TensorData},
};
use rand::{SeedableRng, rngs::StdRng};
use thrust_rl::{
    env::{Environment, cartpole::CartPole},
    policy::q_network::QNetworkBurn,
    train::{
        dqn::{DQNConfig, DQNTrainerBurn},
        optimizer::BurnOptimizer,
    },
};
```

The `select_action` closure follows the same pattern as in `train_cartpole_dqn.rs`: build a `[1, obs_dim]` tensor, call `q.forward(o_t)`, argmax manually. Include a prose callout explaining *why* the closure exists (the trainer is generic over `Q`, so it can't call `.forward()` internally without the caller spelling out the type).

Polyak vs hard copy: the tutorial should set `.soft_update_tau(0.005)` and explain what it does (blend `θ_target ← τ·θ_online + (1−τ)·θ_target` every step vs hard copy every `target_update_interval` steps). Note that `DQNConfig::soft_update_tau` is `None` by default (hard copy mode); the examples explicitly set it to `0.005`.

### Step 2 — Update `src/tutorials.rs`

Append after the `tutorial_02_cartpole_ppo` module. Also update the module-level doc comment's "Landed tutorials" bullet list to include tutorial 3. The file currently ends at line ~44; the new module is ~7 lines.

### Step 3 — Update `docs/tutorials/README.md`

Find the row containing `| 3 |` and change `🚧 Planned ([#309](https://github.com/rjwalters/thrust/issues/309))` to `✅ Landed`. Update the link in the "Tutorial" column cell to point to `[Off-policy training with DQN](03-dqn.md)`.

### Scope guard

This issue is a single Builder PR. Estimated size: ~250–350 lines of new/modified content across three files. Do not add a fourth environment or a prioritized-replay variant — those belong in a separate follow-up if desired. Do not modify any `src/train/dqn/` or `src/policy/q_network.rs` files — the DQN API is stable and the tutorial must use it as-is.

---

## Affected Files

- `docs/tutorials/03-dqn.md` — new file; the tutorial document and its doc-tested `rust` block
- `src/tutorials.rs` — add `tutorial_03_dqn` submodule and update module-level doc comment bullet list
- `docs/tutorials/README.md` — update tutorial 3 status row from "🚧 Planned" to "✅ Landed"

All three files verified against `origin/main`: `src/tutorials.rs` and `docs/tutorials/README.md` exist on `origin/main`; `docs/tutorials/03-dqn.md` is a new file (not a pre-existing uncommitted path — confirmed absent from both origin/main and the local working tree).

---

## Test Plan

- [ ] Manual verification: run `cargo test --features training --doc tutorials::tutorial_03_dqn` in the repo root; confirm the doc-test compiles and runs without error.
- [ ] CI gate: `cargo test --features training` passes (includes all doc-tests in `src/tutorials.rs`).
- [ ] Docs gate: `RUSTDOCFLAGS=-D warnings cargo doc --features training` produces no warnings from the new module.
- [ ] Clippy: `cargo clippy --features training -- -D warnings` is clean.
- [ ] Format: `cargo fmt --check` passes.
- [ ] Full run sanity: `cargo run --example train_cartpole_dqn --features training --release` still runs (no regression in the working examples).
- [ ] Tutorial README: confirm `docs/tutorials/README.md` status row 3 shows `✅ Landed` with a working link to `03-dqn.md`.

## Dependencies

- [x] #303: Tutorial infrastructure (`src/tutorials.rs` doc-test mechanism, tutorials 1–2 as style references) — already merged and on `origin/main`.
